fix: shutdown and shutdown notification issues [MTT-7791] [MTT-7540]

NoelStephensUnity opened the pull request:
This PR includes several fixes to the NetworkManager shutdown sequence where:

- A client disconnected by a server-host would not receive a notification
- A server-host could shutdown during a relay connection but periodically the transport disconnect message to any connected clients could be dropped.
- A host could disconnect its local client but remain running as a server
- OnClientDisconnectedCallback was not being invoked under certain conditions when disconnected.
- OnClientDisconnectedCallback was not returning the assigned client identifier on the client-side when invoked.
- If a host or server shutdown while a client owned NetworkObjects (other than the player) it would throw an exception.
- If you modify a NetworkVariable or NetworkList during a NetworkManager shutdown an exception was thrown.

[MTT-7791](https://jira.unity3d.com/browse/MTT-7791)
[MTT-7540](https://jira.unity3d.com/browse/MTT-7540)

fix: #2759
fix: #2705
fix: #2699
fix: #2389
fix: #2064
fix: #1880

## Changelog

- Fixed: Issue where a client disconnected by a server-host would not receive a local notification.
- Fixed: Issue where a server-host could shutdown during a relay connection but periodically the transport disconnect message sent to any connected clients could be dropped.
- Fixed: Issue where a host could disconnect its local client but remain running as a server.
- Fixed: Issue where `OnClientDisconnectedCallback` was not being invoked under certain conditions.
- Fixed: Issue where `OnClientDisconnectedCallback` was always returning 0 as the client identifier.
- Fixed: Issue where if a host or server shutdown while a client owned NetworkObjects (other than the player) it would throw an exception.
- Fixed: Issue where setting values on a `NetworkVariable` or `NetworkList` within `OnNetworkDespawn` during a shutdown sequence would throw an exception.
- Changed: The server or host shutdown so it will now perform a "soft shutdown" when `NetworkManager.Shutdown` is invoked. This will send a disconnect notification to all connected clients and the server-host will wait for all connected clients to disconnect or timeout after a 5 second period before completing the shutdown process.
- Changed: `OnClientDisconnectedCallback` will now return the assigned client identifier on the local client side if the client was approved and assigned one prior to being disconnected.


## Testing and Documentation

- Includes integration test updates.
- Updates to public documentation: [PR-1149](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/1149)

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
fix

Just return from OnClientDisconnectFromServer when the server is shutting down as there is no point in doing ownership cleanup or client disconnected notification messages at that point.

NoelStephensUnity committed:
Fix

Approaching this fix a different way since we still should send out client disconnect notifications and such.

NoelStephensUnity committed:
fix

Server or host was not sending disconnect when it shutdown.

NoelStephensUnity committed:
Merge branch 'develop' into fix/client-owned-objects-cause-exception-upon-host-server-shutdown

NoelStephensUnity committed:
update

When a client disconnects itself or is disconnected by the server, it now returns its client identifier.

Server or host shutdown sequence now includes a "soft shutdown" where the server-host will send out disconnection messages to all clients with the reason for the client being disconnected.

NoelStephensUnity committed:
update

the test project exit button script is now a bit more intelligent with how it handles exiting and now logs the reason for disconnection.

NoelStephensUnity committed:
fix - testproject

fixing exceptions when certain missing properties are not set (primarily buttons or toggles)

NoelStephensUnity committed:
test

updating tests for changes.

NoelStephensUnity committed:
fix

Don't allow a host or server to disconnect its local client.

NoelStephensUnity committed:
update

Adding changelog entries

NoelStephensUnity committed:
style

removing unused namespace.

NoelStephensUnity committed:
fix

Fixes issue with NetworkVariable and NetworkList throwing exceptions upon shutdown.

NoelStephensUnity committed:
Test

Validates that setting NetworkVariables or NetworkList values during OnNetworkDespawn does not throw an exception.

NoelStephensUnity committed:
update

adding change log entry

NoelStephensUnity committed:
style

fixing warning message about setting network variable/list values during a shutdown.

NoelStephensUnity committed:
test

Removing unrequired null check

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -347,7 +347,7 @@ internal ulong TransportIdCleanUp(ulong transportId)
             if (!LocalClient.IsServer && !TransportIdToClientIdMap.ContainsKey(transportId))
             {
-                return 0;
+                return NetworkManager.LocalClientId;
```
I am curious... if we are in the state as indicated by the comment above, how would LocalClientId ever be set? Wouldn't it still just be 0? I'm not sure returning a default-initialized variable that was never set is actually valid here.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -925,6 +935,7 @@ internal void OnClientDisconnectFromServer(ulong clientId)
                         }
                     }
                     else
+                    if (!NetworkManager.ShutdownInProgress)
```
Any particular reason you're putting all these ifs on their own line instead of using the standard `else if`?

NoelStephensUnity replied:
It is the standard way to do that, but can make it all one line.

NoelStephensUnity replied:
I am mistaken... I have Rust in my brain.
"else if" is the correct way (_my brain went on vacation Wednesday night_)

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -1077,16 +1095,9 @@ internal void Initialize(NetworkManager networkManager)
-            ConnectedClients.Clear();
-            ConnectedClientIds.Clear();
-            ConnectedClientsList.Clear();
+
```
Not a critique, just curious why these (and ProcessSendQueues) needed to be moved, just so I can understand the changes you made better.

NoelStephensUnity replied:
Ahh... yeah this is because the `DisconnectRemoteClient(clientId)` invocation below on line 1132 can invoke OnClientDisconnectFromServer which invokes `DisconnectRemoteClient`:
```
        internal void DisconnectRemoteClient(ulong clientId)
        {
            MessageManager.ProcessSendQueues();
            OnClientDisconnectFromServer(clientId);
        }
```
It processes the send queue prior to invoking `OnClientDisconnectFromServer(clientId);`. However, on line 995 (within `OnClientDisconnectFromServer`) the `ClientDisconnectedMessage` is being sent to any remaining connected clients. This means the last client to be disconnected would not get the `ClientDisconnectedMessage` if a host or server was shutting down.

I didn't want to change the order of operations in `DisconnectRemoteClient` but wanted to make sure the last `ClientDisconnectedMessage` was sent out.
_(it is sort of a mute point...but I guess I was just in the "make sure any last messages are sent" mode)_

Thinking about it... the last client would get the last message... so I probably don't need to do that... and since I am sending disconnect messages prior to even shutting down... that probably doesn't even need to be invoked any longer.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/NetworkVariable/NetworkVariableBase.cs:
```diff
@@ -100,7 +100,16 @@ protected void MarkNetworkBehaviourDirty()
+            {
+                if (m_NetworkBehaviour.NetworkManager.LogLevel <= LogLevel.Developer)
+                {
+                    Debug.LogWarning($"NetworkVariable is written to during the NetworkManager shutdown! " +
```
The wording of this message feels a little passive aggressive to me. "Are you doing X? If so, try not to do X." I feel like the second sentence is implied by the question and can be left off.

NoelStephensUnity committed:
style

remove the cr/lf after some else if statments

ShadauxCat committed:
Minor formatting changes.

NoelStephensUnity committed:
update

Adjusting warning message language.
Adjusting when the send queue is processed.

NoelStephensUnity committed:
Merge branch 'fix/client-owned-objects-cause-exception-upon-host-server-shutdown' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/client-owned-objects-cause-exception-upon-host-server-shutdown

NoelStephensUnity committed:
Merge branch 'develop' into fix/client-owned-objects-cause-exception-upon-host-server-shutdown

NoelStephensUnity committed:
update

reverting back moving the MessageManager?.ProcessSendQueues(); within Shutdown and removing the migration of that out of OnClientDisconnectFromServer.

netcode-ci-service committed:
Merge develop into fix/client-owned-objects-cause-exception-upon-host-server-shutdown
